internal/fwserver/server_listresources.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package fwserver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/internal/fwschema"
+	"github.com/hashicorp/terraform-plugin-framework/internal/logging"
+	"github.com/hashicorp/terraform-plugin-framework/list"
+	"github.com/hashicorp/terraform-plugin-framework/provider"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+)
+
+// ListResourceFuncs returns a map of ListResource functions. The results are
+// cached on first use.
+func (s *Server) ListResourceFuncs(ctx context.Context) (map[string]func() list.ListResource, diag.Diagnostics) {
+	provider, ok := s.Provider.(provider.ProviderWithListResources)
+	if !ok {
+		return nil, nil
+	}
+
+	logging.FrameworkTrace(ctx, "Checking ListResourceFuncs lock")
+	s.listResourceFuncsMutex.Lock()
+	defer s.listResourceFuncsMutex.Unlock()
+
+	if s.listResourceFuncs != nil {
+		return s.listResourceFuncs, s.listResourceFuncsDiags
+	}
+
+	providerTypeName := s.ProviderTypeName(ctx)
+	s.listResourceFuncs = make(map[string]func() list.ListResource)
+
+	logging.FrameworkTrace(ctx, "Calling provider defined ListResources")
+	listResourceFuncSlice := provider.ListResources(ctx)
+	logging.FrameworkTrace(ctx, "Called provider defined ListResources")
+
+	for _, listResourceFunc := range listResourceFuncSlice {
+		listResource := listResourceFunc()
+
+		metadataReq := resource.MetadataRequest{
+			ProviderTypeName: providerTypeName,
+		}
+		metadataResp := resource.MetadataResponse{}
+		listResource.Metadata(ctx, metadataReq, &metadataResp)
+
+		typeName := metadataResp.TypeName
+		if typeName == "" {
+			s.listResourceFuncsDiags.AddError(
+				"ListResource Type Name Missing",
+				fmt.Sprintf("The %T ListResource returned an empty string from the Metadata method. ", listResource)+
+					"This is always an issue with the provider and should be reported to the provider developers.",
+			)
+			continue
+		}
+
+		logging.FrameworkTrace(ctx, "Found resource type", map[string]interface{}{logging.KeyListResourceType: typeName}) // TODO: y?
+
+		if _, ok := s.listResourceFuncs[typeName]; ok {
+			s.listResourceFuncsDiags.AddError(
+				"Duplicate ListResource Type Defined",
+				fmt.Sprintf("The %s ListResource type name was returned for multiple list resources. ", typeName)+
+					"ListResource type names must be unique. "+
+					"This is always an issue with the provider and should be reported to the provider developers.",
+			)
+			continue
+		}
+
+		if _, ok := s.resourceFuncs[typeName]; !ok {
```

Comment:
Is `s.resourceFuncs` guaranteed to be populated at this point?

Reply:
Maybe we change this to `s.ResourceFuncs[typeName]` ...